File: programmers/42577.cpp
// 9:56
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

bool solution(vector<string> phone_book) {
    bool answer = true;

    unordered_map<string, bool> num_table;

    for (auto &num : phone_book) num_table.insert({num, true});
    
    for (int i=0; i<phone_book.size(); ++i) {
        for (int j=1; j<=20; ++j) {
            if (phone_book[i].size() <= j) break;
            if (num_table.find(phone_book[i].substr(0, j)) != num_table.end()) return false;
        }
    }
    return answer;
}

int main() {
    vector<string> test = {"1195524421", "97674223", "119"};

    bool answ = solution(test);

    cout<<answ<<endl;

}

File: programmers/42577.py
def solution(phone_book):
    answer = True
    result_set = set()
    for num in phone_book:
        result_set.add(num)
    for num in phone_book:
        for i in range(1,len(num)):
            if num[0:i] in result_set:
                return False
    return answer

    # 왜 해시...? key값 존재유무만 확인하는데 굳이?